### ISSUE TITLE
JITServer exception/assert: Arity mismatch: Expected 1 args to unpack but got 9-tuple

### ISSUE THREAD

**Issue by a7ehuo** [eclipse/openj9#6729]
It looks to me if this exception or assert is generated at the end of the compilation, it could be false positive. 

Two issues here: 
1. The error message might not be accurate. The code checks the received message size against the number of args to be expected. It'd be more accurate to say `"Received " + std::to_string(message->data_size()) + " args to unpack but expect " + std::to_string(sizeof...(Args)) + "-tuple");`

```C++
   template <typename... Args>
   std::tuple<Args...> getArgs(const AnyData *message)
      {
      if (sizeof...(Args) != message->data_size())
         throw StreamArityMismatch("Expected " + std::to_string(message->data_size()) + " args to unpack but got " + std::to_string(sizeof...(Args)) + "-tuple");
      return GetArgs<Args...>::getArgs(message, 0);
      }
```

2. It could be a false positive for the end of compilation messages. If there is a compilation failure on the server side. `finishCompilation()` with `MessageType::compilationCode` will not send tuple back to the client but send only the statusCode. 

```
=== Server === 
#FAILURE:  <EARLY TRANSLATION FAILURE: compilation aborted>finishCompilation: statusCode 55, arguments size 0

=== Client ===
#9  <signal handler called>
#10 __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#11 0x00007ffff73e8801 in __GI_abort () at abort.c:79
#12 0x00007ffff73d839a in __assert_fail_base (fmt=0x7ffff755f7d8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x7fffef776755 "0", 
    file=file@entry=0x7fffef776700 "/root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/omr/compiler/infra/Assert.cpp", line=line@entry=69, 
    function=function@entry=0x7fffef7767d0 <TR::trap()::__PRETTY_FUNCTION__> "void TR::trap()") at assert.c:92
#13 0x00007ffff73d8412 in __GI___assert_fail (assertion=0x7fffef776755 "0", 
    file=0x7fffef776700 "/root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/omr/compiler/infra/Assert.cpp", line=69, 
    function=0x7fffef7767d0 <TR::trap()::__PRETTY_FUNCTION__> "void TR::trap()") at assert.c:101
#14 0x00007fffeefb3d45 in TR::trap () at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/omr/compiler/infra/Assert.cpp:69
#15 0x00007fffeefb41e7 in TR::fatal_assertion (file=0x7fffef657438 "/root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/net/StreamExceptions.hpp", line=127, 
    condition=0x7fffef657117 "false", format=0x7fffef6574c4 "Arity mismatch: %s") at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/omr/compiler/infra/Assert.cpp:160
#16 0x00007fffeefb411c in TR::assertion (file=0x7fffef657438 "/root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/net/StreamExceptions.hpp", line=127, 
    condition=0x7fffef657117 "false", format=0x7fffef6574c4 "Arity mismatch: %s") at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/omr/compiler/infra/Assert.cpp:151
#17 0x00007fffee9a5aa6 in JITServer::StreamArityMismatch::StreamArityMismatch (this=0x7fffd0029870, message=..., __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/net/StreamExceptions.hpp:127
#18 0x00007fffeebf6b86 in JITServer::getArgs<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::tuple<std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_ResolvedMethod*, std::allocator<TR_ResolvedMethod*> >, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<std::tuple<VirtualGuardInfoForCHTable, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<VirtualGuardInfoForCHTable, std::allocator<VirtualGuardInfoForCHTable> > >, std::allocator<std::tuple<VirtualGuardInfoForCHTable, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<VirtualGuardInfoForCHTable, std::allocator<VirtualGuardInfoForCHTable> > > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, unsigned char*>, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<TR_ResolvedJ9Method*, std::allocator<TR_ResolvedJ9Method*> >, TR_OptimizationPlan> (message=0x7fffd000fee0) at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/net/ProtobufTypeConvert.hpp:283
#19 0x00007fffeebe5f9e in JITServer::ClientStream::getRecvData<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::tuple<std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_ResolvedMethod*, std::allocator<TR_ResolvedMethod*> >, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<std::tuple<VirtualGuardInfoForCHTable, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<VirtualGuardInfoForCHTable, std::allocator<VirtualGuardInfoForCHTable> > >, std::allocator<std::tuple<VirtualGuardInfoForCHTable, std::vector<TR_VirtualGuardSite, std::allocator<TR_VirtualGuardSite> >, std::vector<VirtualGuardInfoForCHTable, std::allocator<VirtualGuardInfoForCHTable> > > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocat--Type <RET> for more, q to quit, c to continue without paging--
or<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, unsigned char*>, std::vector<TR_OpaqueClassBlock*, std::allocator<TR_OpaqueClassBlock*> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<TR_ResolvedJ9Method*, std::allocator<TR_ResolvedJ9Method*> >, TR_OptimizationPlan> (this=0x7fffec449af8) at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/net/ClientStream.hpp:138
#20 0x00007fffeebcdc5d in remoteCompile (vmThread=0x637e00, compiler=0x7fff82000000, compilee=0x7fffeb1c9938, method=0x66bc00, details=..., compInfoPT=0x7fffec3e1970)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/JITaaSCompilationThread.cpp:2674
#21 0x00007fffee9c2846 in TR::CompilationInfoPerThreadBase::compile (this=0x7fffec3e1970, vmThread=0x637e00, compiler=0x7fff82000000, compilee=0x7fffeb1c9938, vm=..., 
    optimizationPlan=0x7fffec491870, scratchSegmentProvider=...) at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:8901
#22 0x00007fffee9c0a89 in TR::CompilationInfoPerThreadBase::wrappedCompile (portLib=0x7ffff63eb3a0 <j9portLibrary>, opaqueParameters=0x7fffeb1c96b0)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:8403
#23 0x00007ffff5578d93 in omrsig_protect (portLibrary=0x7ffff63eb3a0 <j9portLibrary>, 
    fn=0x7fffee9be34a <TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary*, void*)>, fn_arg=0x7fffeb1c96b0, 
    handler=0x7fffee9a93a8 <jitSignalHandler(J9PortLibrary*, U_32, void*, void*)>, handler_arg=0x637e00, flags=249, result=0x7fffeb1c9680) at ../../omr/port/unix/omrsignal.c:416
#24 0x00007fffee9bded5 in TR::CompilationInfoPerThreadBase::compile (this=0x7fffec3e1970, vmThread=0x637e00, entry=0x7ffff045c1c0, scratchSegmentProvider=...)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:7593
#25 0x00007fffee9b4689 in TR::CompilationInfoPerThread::processEntry (this=0x7fffec3e1970, entry=..., scratchSegmentProvider=...)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:4141
#26 0x00007fffee9b388d in TR::CompilationInfoPerThread::processEntries (this=0x7fffec3e1970)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:3847
#27 0x00007fffee9b312e in TR::CompilationInfoPerThread::run (this=0x7fffec3e1970)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:3717
#28 0x00007fffee9b2eea in protectedCompilationThreadProc (compInfoPT=0x7fffec3e1970)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:3660
#29 0x00007ffff5578d93 in omrsig_protect (portLibrary=0x7ffff63eb3a0 <j9portLibrary>, 
    fn=0x7fffee9b2d04 <protectedCompilationThreadProc(J9PortLibrary*, TR::CompilationInfoPerThread*)>, fn_arg=0x7fffec3e1970, handler=0x7ffff5c92400 <structuredSignalHandler>, 
    handler_arg=0x637e00, flags=506, result=0x7fffeb1cad98) at ../../omr/port/unix/omrsignal.c:416
#30 0x00007fffee9b2ca8 in compilationThreadProc (entryarg=0x7fffec3e1970) at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/CompilationThread.cpp:3565
#31 0x00007ffff59e62c6 in thread_wrapper (arg=0x7ffff00d1620) at ../omr/thread/common/omrthread.c:1605
#32 0x00007ffff7bbd6db in start_thread (arg=0x7fffeb1cd700) at pthread_create.c:463
#33 0x00007ffff74c988f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 

(gdb) f 20
#20 0x00007fffeebcdc5d in remoteCompile (vmThread=0x637e00, compiler=0x7fff82000000, compilee=0x7fffeb1c9938, method=0x66bc00, details=..., compInfoPT=0x7fffec3e1970)
    at /root/home/ahuo/src/openj9-openjdk-jdk8/openj9/runtime/compiler/control/JITaaSCompilationThread.cpp:2674
2674	                                      std::string, std::string, std::vector<TR_ResolvedJ9Method*>, TR_OptimizationPlan>();
(gdb) list -
2669	         comp->failCompilation<TR::CompilationInterrupted>("Compilation interrupted in remoteCompile");
2670	         }
2671	
2672	      while(!handleServerMessage(client, compiler->fej9vm()));
2673	      auto recv = client->getRecvData<uint32_t, std::string, std::string, CHTableCommitData, std::vector<TR_OpaqueClassBlock*>,
2674	                                      std::string, std::string, std::vector<TR_ResolvedJ9Method*>, TR_OptimizationPlan>();
```

**Comment by mpirvu** [comment]
So the error message is wrong: it expect 9 arguments, but it only got one. Nice find!
There are two possible fixes:
1) Always send 9 arguments, even in case of error. The arguments that are not needed could be empty
2) Create a different type of message that is sent when the compilation fails at the server. The client will only try to read the error code in this case.

I feel that solution 2 is cleaner.

**Comment by harryyu1994** [comment]
solution 1 is much easier to implement, but solution 2 is probably easier for people to understand.

**Comment by a7ehuo** [comment]
I'll implement solution 2.

**Comment by harryyu1994** [comment]
FYI, apparently this fixed a whole bunch of asserts I was getting. I made a patch (solution 1) for my own code it looks like this fixed the asserts I was getting in #6687.

**Comment by a7ehuo** [comment]
This issue is addressed in #6733 which has been merged. Close
